src/OrleansProviders/Streams/Common/PooledCache/ObjectPool.cs

```diff
@@ -0,0 +1,44 @@
+﻿
+using System;
+using System.Collections.Generic;
+
+namespace Orleans.Providers.Streams.Common
+{
+    /// <summary>
+    /// Simple object pool that uses a stack to store available objects.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ObjectPool<T> : IObjectPool<T>
+        where T : PooledResource<T>
+    {
+        private const int DefaultPoolCapacity = 1 << 10; // 1k
+        private readonly Stack<T> pool;
+        private readonly Func<IObjectPool<T>, T> factoryFunc;
+
+        public ObjectPool(Func<IObjectPool<T>, T> factoryFunc, int initialCapacity = DefaultPoolCapacity)
+        {
+            if (factoryFunc == null)
+            {
+                throw new ArgumentNullException("factoryFunc");
+            }
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException("initialCapacity");
+            }
+            this.factoryFunc = factoryFunc;
+            pool = new Stack<T>(initialCapacity);
+        }
+
+        public virtual T Allocate()
+        {
+            return pool.Count != 0
+                ? pool.Pop()
+                : factoryFunc(this);
+        }
+
+        public void Free(T resource)
+        {
+            pool.Push(resource);
```

Comment:
>"not allowing the GC to even collect the PooledResource... is that by design"

Yes.  We allocate relatively few large objects that contain few object references and never free them.  The fact that there are relatively few objects with few references makes it cheap for garbage collection to check to see if they need collected, and the fact that we never free them means that they will eventually settle in the garbage collector and no longer (or rarely) be moved around or trigger other data to be moved during compaction.

Is all of this necessary?  Unclear.  We've not tested the large objects without the object pools to see what the perf hit is, but we know that keeping the objects in pools is cheaper than letting the garbage collection constantly compact and collect them, we just don't know how much cheaper it is.